### ISSUE TITLE
convert ckpt.pt to hugging face model

### ISSUE THREAD

**Issue by hujunchao** [karpathy/llama2.c#321]
Thanks for the amazing job! I have a question that after training, can I convert ckpt.pt to hugging face model? And can I use the LlamaForCausalLM to inference?

**Comment by karpathy** [comment]
In principle absolutely. In practice noone has submitted a PR to export it and I don't personally care as much, so I've been ignoring it :) Would probably accept the PR in the new export.py file that is about to merge.

**Comment by xenova** [comment]
@hujunchao If you're looking for transformers-compatible versions of the 15M, 42M, or 110M models, you can use mine on the [HF hub](https://huggingface.co/models?sort=trending&search=Xenova%2Fllama2.c-stories).

If you want to convert your own, here are the steps I took:
1. Download the pt files from the hub. For example, the 15M version:
    ```sh
    !wget https://huggingface.co/karpathy/tinyllamas/resolve/main/stories15M.bin
    ```
2. Update weight names:
    ```py
    import torch
    import os
    
    version = 15 # 15, 42, or 110
    data = torch.load(f"stories{version}M.pt")
    
    renamed_data = {}
    for k, v in data["model"].items():
        k = k.replace("layers", "model.layers")
        k = k.replace(".feed_forward.w1", ".mlp.gate_proj")
        k = k.replace(".feed_forward.w2", ".mlp.down_proj")
        k = k.replace(".feed_forward.w3", ".mlp.up_proj")
        k = k.replace(".attention.wq", ".self_attn.q_proj")
        k = k.replace(".attention.wk", ".self_attn.k_proj")
        k = k.replace(".attention.wv", ".self_attn.v_proj")
        k = k.replace(".attention.wo", ".self_attn.o_proj")
        k = k.replace("norm.weight", "model.norm.weight")
        k = k.replace("output.weight", "lm_head.weight")
        k = k.replace("ffn_model.norm", "post_attention_layernorm")
        k = k.replace("attention_model.norm", "input_layernorm")
        k = k.replace("tok_embeddings.weight", "model.embed_tokens.weight")
        renamed_data[k] = v
    
    output_folder = f'llama2.c-stories{version}M'
    os.makedirs(output_folder, exist_ok=True)
    torch.save(renamed_data, f"{output_folder}/pytorch_model.bin")
    ```
3. Create and add a config.json to your model folder. See [mine](https://huggingface.co/Xenova/llama2.c-stories15M/blob/main/config.json) for an example. All you need to do is update the parameters (e.g., `hidden_size`, `num_hidden_layers` and so on).
4. Copy the [tokenizer.json](https://huggingface.co/Xenova/llama2.c-stories15M/resolve/main/tokenizer.json) and [tokenizer_config.json](https://huggingface.co/Xenova/llama2.c-stories15M/resolve/main/tokenizer_config.json) files from my example. They are the same as all the other llama models.
5. Run in transformers:
    ```py
    from transformers import AutoModelForCausalLM, AutoTokenizer
    
    tokenizer = AutoTokenizer.from_pretrained(output_folder) 
    model = AutoModelForCausalLM.from_pretrained(output_folder)
    
    inputs = tokenizer.encode("", return_tensors="pt")
    outputs = model.generate(inputs)
    print(tokenizer.decode(outputs[0]))
    # <s> Once upon a time, there was a big bear named Benny was a very big bear.
    ```

**Comment by photomz** [comment]
Thanks @xenova , just what I needed! Could we merge this into `export.py` as a PR? I'll add PyTests and sanity checks, of course

**Comment by xenova** [comment]
Sure thing!

**Comment by photomz** [comment]
Hey @xenova , I'm trying to replicate your steps in a notebook, but I notice your HuggingFace model `Xenova/llama2.c-stories15M` yields different generations than `sample.py`:

```py
from transformers import AutoModelForCausalLM, AutoTokenizer

output_folder = 'Xenova/llama2.c-stories15M'

tokenizer = AutoTokenizer.from_pretrained(output_folder) 
model = AutoModelForCausalLM.from_pretrained(output_folder)

inputs = tokenizer.encode("Lily was", return_tensors="pt")
outputs = model.generate(inputs, max_length=256)
print(tokenizer.decode(outputs[0]))

# <s> Lily was a squir toy of a girl who liked toy. She was a girl. She was her name her favorite friend. She was a boy. She was three-ar. She had a bit. She was aunt, too. She was called Lily, butter. She was a little girl who was three-d-die was three.
# One day, and she was Lily. She was called her mommy. She was called her age.
# One day, and she was called her first. She was called Lily'dusted. She was a bitter. She was a little girl.
```

Compared to `sample.py`, both temperature=0, top_p=1:

```
❯ python sample.py --checkpoint=stories15M.pt --start="Lily was"
Overriding: checkpoint = stories15M.pt
Overriding: start = Lily was
Lily was a new girl in the classroom. She liked to play with her friends, but she had a job to do. She had to draw on the whiteboard with chalk. She had to write letters and numbers with a friend.
"Time to work," her teacher said. "You can use the chalk to write."
Lily did not like to work. She liked to draw. She complained to her teacher. "Red means be nice to me," Lily said.
```

The TinyStories paper always reported grammatically correct sentences, so the HF model repeating itself seems a conversion issue.

Perhaps it's related to weight tying (I thought this only affect ONNX but maybe not)? Insights appreciated 😄

**Comment by pesvut** [comment]
Didn't see this before, but I have submitted PR https://github.com/karpathy/llama2.c/pull/395 trying to address the same thing

**Comment by xenova** [comment]
Hi @pesvut! Did you manage to fix the [issues](https://github.com/karpathy/llama2.c/issues/321#issuecomment-1704665888) highlighted by @photomz by any chance? This has also stumped me with my ONNX exports.

**Comment by photomz** [comment]
+1, would love to hear if anyone finds a fix for this.

**Comment by pesvut** [comment]
It works fine for me, based on [the models I uploaded to HF hub](https://huggingface.co/models?sort=trending&search=nickypro%2Ftinyllama). I would guess the issue for Xenova was that the key and query matrices needed to be rearranged for the HuggingFace implementation.

Example input:
```
from transformers import AutoModelForCausalLM, AutoTokenizer
model = AutoModelForCausalLM.from_pretrained("nickypro/tinyllama-15M")
tokenizer = AutoTokenizer.from_pretrained("nickypro/tinyllama-15M")
inputs = tokenizer.encode("Lily was", return_tensors="pt")
outputs = model.generate(inputs, max_length=256)
print(tokenizer.decode(outputs[0]))
```

Example output:
```
<s> Lily was a little girl who liked to play with her toys. She had a doll, a ball, a car and a bear. She liked to make them talk and have fun. But sometimes she was naughty and did not listen to her mom and dad.
One day, Lily was playing with her bear in the living room. She saw a big box on the table. She wanted to see what was inside. She climbed on a chair and reached for the box. But the box was too heavy and she lost her balance. She fell off the chair and the box fell on her head. She screamed and cried.
Her mom and dad heard her scream and ran to the living room. They saw Lily on the floor and the box on the floor. They were very angry and worried. They picked up Lily and hugged her. They checked if she was hurt. She was not hurt, but she was very scared and sorry.
Her mom and dad took her to the hospital. The doctor said she had a big bump on her head and a big bruise on her arm. He said she had to stay in bed for a long time and she would be okay
```

**Comment by xenova** [comment]
> I would guess the issue for Xenova was that the key and query matrices needed to be rearranged for the HuggingFace implementation.

Thanks for that!

The last thing which I think needs solving is the issue with model size (see [here](https://github.com/karpathy/llama2.c/issues/378)) - which looks to be an issue with HF transformers.
```py
from transformers import AutoModelForCausalLM
model = AutoModelForCausalLM.from_pretrained("nickypro/tinyllama-15M")
print(sum([v.numel() for v in model.parameters()])) # outputs 24407712
```
and doing `.save_pretrained()` gives a ~93MB model (as opposed to a ~60MB model). It looks like the tied weights are being saved twice. I'll look into it more.

**Comment by pesvut** [comment]
I think this is just because for some reason, with language models, it is standard convention to not count the token embedding parameters when totalling the number of parameters. For example, see [this discussion](https://huggingface.co/EleutherAI/pythia-70m#naming-convention-and-parameter-count) by EleutherAI for their pythia models.

If you don't count the input embeddings, you get the stated 15M parameters:

```
from transformers import AutoModelForCausalLM
model = AutoModelForCausalLM.from_pretrained(args.model)

# count parameters excluding embed_tokens.weight
total_params = 0
for k, v in model.state_dict().items():
    if k == "model.embed_tokens.weight":
        continue
    total_params += v.numel()
print(total_params/1e6) # outputs 15.191712 M
```

**Comment by xenova** [comment]
Although, the original checkpoints (e.g., like [these](https://huggingface.co/karpathy/tinyllamas/tree/main) on the HF hub):
![image](https://github.com/karpathy/llama2.c/assets/26504141/21d6f9cb-775b-4de8-9e8a-cdbd57fac832)

seem to be the correct size? So, there is indeed duplication, right?

**Comment by pesvut** [comment]
Ok, I was briefly confused why the size is smaller on the stories15M. I load:

```
>>> params = torch.load("orig_models/stories15M.pt")
>>> print(sum([v.numel() for k, v in params['model'].items()]))
24407712
>>> print([v.dtype==torch.float32  for k, v in params['model'].items()])
[True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
```

Seems that in (L242 of model.py)[https://github.com/karpathy/llama2.c/blob/766a30bc6e9a1c69ce007bb69caabf4c6062f0e9/model.py#L224] they are doing weight tying.

```
# share the unembedding parameters with the embedding parameters
self.tok_embeddings.weight = self.output.weight # https://paperswithcode.com/method/weight-tying
```

It used to be common in GPT2 and such to have the output embeddings be the transpose of the input embeddings (ie: `lm_head.weight = token_embed.weight.T`) (This is likely why recent models sometimes forget to count lm_head weights).

More recently people have changed to using separate input and output embedding weights due to better performance. Looking at [`LlamaForCausalLM`](https://github.com/huggingface/transformers/blob/0a55d9f7376f72ad3ff296d4249840021b03bcc4/src/transformers/models/llama/modeling_llama.py#L742C7-L742C23), I think just modify my export code so that
`lm_head.weight = token_embed.weight` to make it the same as in older models.

**Comment by pesvut** [comment]
Ok, I have pushed the changes for tied weights, and made the models with fp16 and fp32 [available on the HuggingFace hub](https://huggingface.co/models?sort=trending&search=nickypro%2Ftinyllama)

**Comment by dopu2k16** [comment]
Hey @pesvut, I have trained using the mentioned configuration for 15 model and when I use >>> sum([v.numel() for k,v in data["model"].items()])
6551712
I am 6.55M parameters, may you tell me exactly how to count the total parameters, I have trained 15M, 42M, 110M as well keeping multiple_of=32 using cutom tokenizer with different vocab size.

**Comment by win10ogod** [comment]
can I convert hugging face to ckpt.pt model?

**Comment by SpaceCowboy850** [comment]
I'm seeing similar behavior to photomz and xenova...but it's odd and subtle. 
 
I have a custom trained model on a variant of tiny stories amongst other things.

If I put in "Lily was" into both sample.py and my hf sample, sample.py seems fine, but HF is a bit off...
Sample.py Output:
`Lily was a curious and adventurous third grader who loved exploring her grandmother's attic. One day, she stumbled upon an old frame made of a strange and mysterious material. She had never seen anything like it before, and her only doubt was whether it was safe to play with.

Lily's grandmother, a wise and kind woman, asked her what she thought about the frame. "I have no idea," she replied, her voice shaking a bit. "It looks very old and mysterious." Lily didn't believe her grandmother and decided to take the frame outside to show it in the sunlight. She hoped it would be priceless, like all the other treasures in her grandmother's attic.

The sun was shining brightly when Lily brought the frame outside to show it to her grandmother. She had no idea what to expect, but her grandmother told her that the frame had many historical and cultural significance. Lily couldn't believe it! She was so eager to learn more about the frame's past that she began to investigate it further. With her grandmother's guidance, she discovered that the frame was from another era, long before Lily's family was settled in their current land. She was amazed at the unexpected friendship that had been formed between her grandmother and the frame's origin. Write a story about a boy who discovers hidden rooms and secret passages in a capital city.` 

HF Output
`Lily was a skilled archer who lived in the high woods. But one day, while she and her friends were out on their usual hunting trips through clear streams that ran down to deep mountain lakes just beyond town's border of hilly hillsides known as "The Forest." They had always heard stories about some foul creatures lurking within these mountains at nightfall when everyone else would be drifting off into sleep or dreaming like spiders do under full moonlight – eavesdropping? No wonder goblins teased them for being so quiet during holy roodnight activities!`

So...both are English, but despite being shorter, the HF Output quickly meanders into run on sentences that go nowhere and quickly lose coherence with the words that came before.

It gets worse when I turn it into "instruct mode" which basically amounts to having a prompt + bos_token (as that is the way I trained this dataset).

`start_ids = enc.encode("Write a story about an archer named Lily competing in an archery tournament.\n", bos=False, eos=False) + [enc.bos_id]`

Sample.py output:
`During the annual archery tournament in the small town of Willow Creek, third-grader Lily was determined to make a name for herself. She had trained for months, honing her skills with her trusty bow and arrow. As the tournament progressed, Lily's opponents began to fall into a routine of impressive prowess, and she wondered if she would stand out again.

Lily's confidence grew with each successful shot, her arrows finding their mark each time. She found herself mentally preparing for the tournament, thinking of each shot as a testament to her skill. Despite the rigorous training, Lily's relentless drive and determination had made her an invaluable player, and she knew she could make the tournament a success.

As the day of the tournament approached, Lily's nerves began to settle, and she felt confident in her abilities. She practiced day and night, her aim steadily improving with each shot. On the day of the tournament, Lily stood tall and proud, her trusty bow and arrow serving as her steady weapon as she scored her first victory of the day. The tournament may have been competitive, but Lily's opponents had fallen far behind, leaving her looking for a victory when all she needed to do was believe in herself and her abilities.`

`prompt = "Write a story about an archer named Lily competing in an archery tournament.\n"
input_ids = tokenizer(prompt, return_tensors="pt", add_special_tokens=False)['input_ids']
input_ids = torch.cat(( input_ids, torch.tensor([[tokenizer.bos_token_id]])), dim=1)
tokenized_input = {'input_ids': input_ids}`

HF Output
`Once upon, there was once little girl who loved to draw pictures with her crayons and stroke the paper just right for each picture she drew on it when playing games or writing poems because that's what friends did! Her love of drawing had grown even stronger so after school soon everything went as planned then more than breakfast before bedtime - all mommy cooked sausages according only my favorite egg dish vegetarian feed 5 days per week but no meat-based meals anything ever again... except now mama smelled them coming out one fannie by accident at home , maybe next we can have baked bean legs good luck ! Yachssn nudpids yujuh"`

So, the HF output is not following instructions at all, and by the end is just speaking gibberish.

I have tried saving the input tokens from sample.py and directly feeding them to the HF model.generate, and it still produces bad output.  I've tried to lockdown the generation of model.generate to more closely match the parameterization from sample.py like so:

tokenized_output = model.generate(**tokenized_input, max_new_tokens=256, repetition_penalty=1.1, do_sample=True, temperature=1.0, top_k=300, pad_token_id=tokenizer.eos_token_id, eos_token_id=tokenizer.eos_token_id)

Something seems subtly off...and I'm not clear how to investigate...I suppose go layer by layer between the two models and see where we start deviating?  I'm not clear how to do that with HF as the interface is a bit murky.  Any ideas?

**Comment by sadransh** [comment]
@xenova is there anyway to convert customized tokenizer to HF one?  I tried LlamaTokenizer, LlamaTokenizerFast, but does not seem to be the right approach.
